Add method to query Git commit info in IDxcVersionInfo

antiagainst opened the pull request:
This will give users a way to know the commit from which a DXC
library/binary was built from, so that they can compare different
libraries/binaries and report bugs easily.

dxc.exe now also prints the commit info using this new interface
method. It uses the following format:

 ` (dev;<commit-count>-<commit-hash>)`

antiagainst commented:
On my local machine, the built `dxc.exe` correctly outputs:

```
Version: dxcompiler.dll: 1.3(dev;1146-8d6d1bc5); dxil.dll: 1.2(10.0.17134.12)
```

marcelolr commented on include/dxc/dxcapi.h:
```diff
@@ -291,6 +291,7 @@ static const UINT32 DxcVersionInfoFlags_Internal = 2; // Internal Validator (non
 struct __declspec(uuid("b04f5b50-2059-4f12-a8ff-a1e0cde1cc7e"))
 IDxcVersionInfo : public IUnknown {
   virtual HRESULT STDMETHODCALLTYPE GetVersion(_Out_ UINT32 *pMajor, _Out_ UINT32 *pMinor) = 0;
+  virtual HRESULT STDMETHODCALLTYPE GetCommitInfo(_Out_ UINT32 *pCommitCount, _Out_ const char **pCommitHash) = 0;
```
This is changing a COM interface, which breaks users & implementers. The pattern to use is to declare an IDxcVersionInfo2 interface derived from IDxcVersionInfo, and QI for that as needed (with the expectation that some clients might implement only the basic version, not the extended one).

antiagainst replied:
SG. I'll create a `IDxcVersionInfo2` for this new method. I think we also need a new UUID for it?

Besides, do you think it's useful to also add a parameter for the commit branch?

antiagainst replied:
Done. I've grabbed the UUID using `uuidgen` by following https://msdn.microsoft.com/de-de/library/windows/desktop/aa373928(v=vs.85).aspx. It seems that the UUID is not tight to the class name in any sense. Is this expected? Am I doing it correctly?

marcelolr replied:
Correct UUIDs aren't tied to names in general (there's fine print around WinRT template/generic interfaces, but none of that applies here). Thanks!

marcelolr commented on tools/clang/tools/dxc/dxc.cpp:
```diff
@@ -1014,11 +1014,14 @@ void DxcContext::GetCompilerVersionInfo(llvm::raw_string_ostream &OS) {
         m_Opts.ExternalFn.empty() ? "dxcompiler.dll" : m_Opts.ExternalFn.data();
     if (SUCCEEDED(CreateInstance(CLSID_DxcCompiler, &VerInfo))) {
       VerInfo->GetVersion(&compilerMajor, &compilerMinor);
+      VerInfo->GetCommitInfo(&commitCount, &commitHash);
```
This is where you'd QI for the new interface to gather the rest of the fields. Also, should probably release the caller's memory once we're done with it.

antiagainst replied:
Done

marcelolr commented on tools/clang/tools/dxcompiler/dxcompilerobj.cpp:
```diff
@@ -53,6 +53,8 @@
 #include "dxillib.h"
 #include <algorithm>
 
+#include "GitCommitInfo.inc" // Auto generated file containing Git commit info
```
We should have the ability to leverage the preprocesseor revert to the prior implementation (no git checks, no commit information available) for cases where this isn't meaningful for downstream implementations, or actively confusing (because, for example, this ends up over in a different git repo altogether). I.e., have a way to ifdef this out.

antiagainst replied:
SG. I'll add a new CMake option to wrap both the CMake configuration and this around.

antiagainst replied:
Done

marcelolr commented on tools/clang/tools/dxcompiler/dxcompilerobj.cpp:
```diff
@@ -905,6 +907,14 @@ class DxcCompiler : public IDxcCompiler2, public IDxcLangExtensions, public IDxc
+    if (pCommitCount == nullptr || pCommitHash == nullptr)
+      return E_INVALIDARG;
+    *pCommitCount = kGitCommitCount;
+    *pCommitHash = kGitCommitHash;
```
Expectation is that the callee allocates memory for out params and caller releases it, so we'll need to make a copy of this.

antiagainst replied:
Done

antiagainst commented:
Hey @marcelolr, sorry for the late reply. Just back from vacation. :) I believe I've addressed all your comments. PTAL.

One thing I'm still not certain is whether we should also return the branch name. If we intended to maintain multiple branches, it could be useful.

marcelolr commented on tools/clang/tools/dxc/dxc.cpp:
```diff
@@ -1017,10 +1017,22 @@ void DxcContext::GetCompilerVersionInfo(llvm::raw_string_ostream &OS) {
     if (SUCCEEDED(CreateInstance(CLSID_DxcCompiler, &VerInfo))) {
       VerInfo->GetVersion(&compilerMajor, &compilerMinor);
+#ifdef SUPPORT_QUERY_GIT_COMMIT_INFO
+      if (SUCCEEDED(CreateInstance(CLSID_DxcCompiler, &VerInfo2)))
```
Doesn't matter very much, but you need not create another compiler object - you can simply QueryInterface on VerInfo to get VerInfo2.

antiagainst replied:
Ah, that's good suggestion. Done.

marcelolr commented on include/dxc/dxcapi.h:
```diff
@@ -298,6 +298,13 @@ IDxcVersionInfo : public IUnknown {
   virtual HRESULT STDMETHODCALLTYPE GetFlags(_Out_ UINT32 *pFlags) = 0;
 };
 
+#ifdef SUPPORT_QUERY_GIT_COMMIT_INFO
```
It's probably OK to not guard the interface declaration - whether the thing is implemented or not is a runtime thing that the client will discover via QI.

antiagainst replied:
Yep. Done.
